### PR TITLE
[sdk/javascript] task: add base value

### PR DESCRIPTION
**Pull request by gimre-xymcity**

### CODE REVIEW

File: sdk/javascript/src/BaseValue.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+	}
+
+	if (value < lowerBound || value > upperBound)
+		throw RangeError(`"value" (${value}) is outside of valid ${8 * byteSize}-bit range`);
+
+	return value;
+};
+
+/**
+ * Represents a base int.
```
Comment (Jaguar0625) [inline sdk/javascript/src/BaseValue.js:28]: integer ?

File: sdk/javascript/src/BaseValue.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+		this.value = check(fixedSize, value, isSigned);
+	}
+
+	/**
+	 * Returns byte size of this object.
+	 * @returns {number} Size in bytes.
+	 */
+	size() {
+		return this.byteSize;
+	}
```
Comment (Jaguar0625) [inline sdk/javascript/src/BaseValue.js:49]: is there reason for this vs just renaming byteSize size?

File: sdk/javascript/test/BaseValue_spec.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
+const { BaseValue } = require('../src/BaseValue');
+const converter = require('../src/utils/converter');
+const { expect } = require('chai');
+
+describe.only('BaseValue', () => {
```
Comment (Jaguar0625) [inline sdk/javascript/test/BaseValue_spec.js:5]: only ^^

File: sdk/javascript/test/BaseValue_spec.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+		static deserialize(arrayInput) {
+			return new MockValue(converter.intFromBytes(arrayInput, MockValue.SIZE, false));
+		}
+	}
+	MockValue.SIZE = 8;
+
+	const canCreateBaseValueFactory = isSigned => (value, size, expectedValue) => {
+		// Act:
+		const val = new BaseValue(size, value, isSigned);
```
Comment (Jaguar0625) [inline sdk/javascript/test/BaseValue_spec.js:21]: val/value =? value/rawValue ?

File: sdk/javascript/test/BaseValue_spec.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+		// Assert:
+		expect(val.size()).to.equal(size);
+		expect(val.value).to.equal(expectedValue);
+	};
+
+	const canCreateUnsignedBaseValue = canCreateBaseValueFactory(false);
+	const canCreateSignedBaseValue = canCreateBaseValueFactory(true);
+
+	it('cannot create 64-bit value with invalid type', () => {
+		// Assert:
```
Comment (Jaguar0625) [inline sdk/javascript/test/BaseValue_spec.js:32]: can drop // Assert when test only contains asserts

File: sdk/javascript/test/BaseValue_spec.js [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+		// Act:
+		const val = new BaseValue(size, value, isSigned);
+
+		// Assert:
+		expect(val.size()).to.equal(size);
+		expect(val.value).to.equal(expectedValue);
+	};
+
+	const canCreateUnsignedBaseValue = canCreateBaseValueFactory(false);
+	const canCreateSignedBaseValue = canCreateBaseValueFactory(true);
```
Comment (Jaguar0625) [inline sdk/javascript/test/BaseValue_spec.js:29]: can't these be declared in the (single) test using each?

File: sdk/javascript/test/BaseValue_spec.js [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+			expect(() => new BaseValue(byteSize, NaN)).to.throw('is not an integer');
+			expect(() => new BaseValue(byteSize, -Infinity)).to.throw('is not an integer');
+			expect(() => new BaseValue(byteSize, Infinity)).to.throw('is not an integer');
+			expect(() => new BaseValue(byteSize, undefined)).to.throw('is not an integer');
+		});
+	});
+
+	it('cannot create unsigned with values outside range', () => {
+		// Assert:
+		expect(() => new BaseValue(1, -1)).to.throw('outside of valid 8-bit range');
```
Comment (Jaguar0625) [inline sdk/javascript/test/BaseValue_spec.js:49]: can you add 3rd test of some value in between for each (could also loop over values)

File: sdk/javascript/test/BaseValue_spec.js [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+	});
+
+	it('can create via converter', () => {
+		// Act:
+		const val = MockValue.deserialize(new Uint8Array([0x23, 0x01, 0xEF, 0xCD, 0xAB, 0x89, 0x67, 0x45]));
+
+		// Assert:
+		expect(val.size()).to.equal(8);
+		expect(val.value).to.equal(DEFAULT_VALUE);
+	});
```
Comment (Jaguar0625) [inline sdk/javascript/test/BaseValue_spec.js:122]: unsure what you're trying to test here ?

File: sdk/javascript/test/BaseValue_spec.js (hunk 3 above)
Comment (gimre-xymcity) [reply sdk/javascript/test/BaseValue_spec.js:5]: 😅

File: sdk/javascript/test/BaseValue_spec.js (hunk 4 above)
Comment (gimre-xymcity) [reply sdk/javascript/test/BaseValue_spec.js:21]: this mimics py, where field name == .value
(and in py tests there's actually val used and `val.value`)

File: sdk/javascript/test/BaseValue_spec.js (hunk 6 above)
Comment (gimre-xymcity) [reply sdk/javascript/test/BaseValue_spec.js:29]: 😅

File: sdk/javascript/src/BaseValue.js (hunk 2 above)
Comment (gimre-xymcity) [reply sdk/javascript/src/BaseValue.js:49]: explained on discord, was supposed to be private,

dropped - both unsure about privates, so nay

File: sdk/javascript/test/BaseValue_spec.js (hunk 7 above)
Comment (gimre-xymcity) [reply sdk/javascript/test/BaseValue_spec.js:49]: not sure if I get you right, but don't you mean next test? `can create unsigned base value`

File: sdk/javascript/test/BaseValue_spec.js (hunk 8 above)
Comment (gimre-xymcity) [reply sdk/javascript/test/BaseValue_spec.js]: so, I do realize the test looks fake, what I want all basevalues to do is to do deserialize like mock value above
(so via `converter.bytesToInt` )

any suggestions?

File: sdk/javascript/test/BaseValue_spec.js (hunk 7 above)
Comment (Jaguar0625) [reply sdk/javascript/test/BaseValue_spec.js:45]: i meant, something like this 
```
		canCreateSignedBaseValue(-0x80, 1, -0x80);
		canCreateSignedBaseValue(0x21, 1, -0x80);
		canCreateSignedBaseValue(0x7F, 1, 0x7F);
```

File: sdk/javascript/test/BaseValue_spec.js (hunk 8 above)
Comment (Jaguar0625) [reply sdk/javascript/test/BaseValue_spec.js]: the real deserialize function will have to be _somewhere_. i'd just test that instead.

File: sdk/javascript/test/BaseValue_spec.js (hunk 7 above)
Comment (gimre-xymcity) [reply sdk/javascript/test/BaseValue_spec.js:45]: oh, the comment was next to _cannot_, and  I couldn't figure out what you mean 😁

**Review by Jaguar0625 — APPROVED** [review]

File: sdk/javascript/test/BaseValue_spec.js [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+		// Assert:
+		// - 8-bit
+		[0, 0x24, 0xFF].forEach(value => canCreateUnsignedBaseValue(value, 1, value));
+
+		// - 16-bit
+		[0, 0x243F, 0xFFFF].forEach(value => canCreateUnsignedBaseValue(value, 2, value));
+
+		// - 32-bit
+		[0, 0x243F6A88, 0xFFFF_FFFF].forEach(value => canCreateUnsignedBaseValue(value, 4, value));
```
Comment (Jaguar0625) [inline sdk/javascript/test/BaseValue_spec.js:56]: missing `_` in second ?

File: sdk/javascript/test/BaseValue_spec.js [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+	it('cannot create signed with values outised range', () => {
+		expect(() => new BaseValue(1, -0x81, true)).to.throw('outside of valid 8-bit range');
+		expect(() => new BaseValue(1, 0x80, true)).to.throw('outside of valid 8-bit range');
+
+		expect(() => new BaseValue(2, -0x8001, true)).to.throw('outside of valid 16-bit range');
+		expect(() => new BaseValue(2, 0x8000, true)).to.throw('outside of valid 16-bit range');
+
+		expect(() => new BaseValue(4, -0x8000_0001, true)).to.throw('outside of valid 32-bit range');
+		expect(() => new BaseValue(4, 0x800_00000, true)).to.throw('outside of valid 32-bit range');
```
Comment (Jaguar0625) [inline sdk/javascript/test/BaseValue_spec.js:71]: 8000_

**Comment by gimre-xymcity** [comment]
merged in 7ab4a1d3666803a469bdecc898a22ce93f4eda00

**Pull request closed without merging**